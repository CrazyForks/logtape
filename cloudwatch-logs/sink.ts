import {
  CloudWatchLogsClient,
  CreateLogStreamCommand,
  type InputLogEvent,
  PutLogEventsCommand,
  ResourceAlreadyExistsException,
} from "@aws-sdk/client-cloudwatch-logs";
import {
  getLogger,
  type LogRecord,
  type Sink,
  type TextFormatter,
} from "@logtape/logtape";
import type { CloudWatchLogsSinkOptions } from "./types.ts";

// AWS CloudWatch Logs PutLogEvents API limits
// See: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html
const MAX_BATCH_SIZE_EVENTS = 10000; // Maximum 10,000 events per batch
const MAX_BATCH_SIZE_BYTES = 1048576; // Maximum batch size: 1 MiB (1,048,576 bytes)
const OVERHEAD_PER_EVENT = 26; // AWS overhead per log event: 26 bytes per event

/**
 * Resolves the log stream name from template.
 * @param logStreamNameTemplate Template for generating stream names
 * @returns Resolved log stream name
 */
function resolveLogStreamName(
  logStreamNameTemplate: string,
): string {
  const now = new Date();
  const year = now.getFullYear().toString();
  const month = (now.getMonth() + 1).toString().padStart(2, "0");
  const day = now.getDate().toString().padStart(2, "0");
  const timestamp = now.getTime().toString();

  return logStreamNameTemplate
    .replace(/\{YYYY\}/g, year)
    .replace(/\{MM\}/g, month)
    .replace(/\{DD\}/g, day)
    .replace(/\{YYYY-MM-DD\}/g, `${year}-${month}-${day}`)
    .replace(/\{timestamp\}/g, timestamp);
}

/**
 * Ensures that the log stream exists, creating it if necessary.
 * @param client CloudWatch Logs client
 * @param logGroupName Log group name
 * @param logStreamName Log stream name
 * @param createdStreams Set to track already created streams
 */
async function ensureLogStreamExists(
  client: CloudWatchLogsClient,
  logGroupName: string,
  logStreamName: string,
  createdStreams: Set<string>,
): Promise<void> {
  const streamKey = `${logGroupName}/${logStreamName}`;

  // If we've already created this stream, skip
  if (createdStreams.has(streamKey)) {
    return;
  }

  try {
    const command = new CreateLogStreamCommand({
      logGroupName,
      logStreamName,
    });

    await client.send(command);
    createdStreams.add(streamKey);
  } catch (error) {
    if (error instanceof ResourceAlreadyExistsException) {
      // Stream already exists, this is fine
      createdStreams.add(streamKey);
    } else {
      // Log stream creation failure to meta logger
      const metaLogger = getLogger(["logtape", "meta", "cloudwatch-logs"]);
      metaLogger.error(
        "Failed to create log stream {logStreamName} in group {logGroupName}: {error}",
        { logStreamName, logGroupName, error },
      );
      // Re-throw other errors
      throw error;
    }
  }
}

/**
 * Gets a CloudWatch Logs sink that sends log records to AWS CloudWatch Logs.
 *
 * @param options Configuration options for the CloudWatch Logs sink.
 * @returns A sink that sends log records to CloudWatch Logs.
 * @since 1.0.0
 */
export function getCloudWatchLogsSink(
  options: CloudWatchLogsSinkOptions,
): Sink & AsyncDisposable {
  const client = options.client ??
    new CloudWatchLogsClient({
      region: options.region ?? "us-east-1",
      credentials: options.credentials,
    });

  const batchSize = Math.min(
    Math.max(options.batchSize ?? 1000, 1),
    MAX_BATCH_SIZE_EVENTS,
  );
  const flushInterval = options.flushInterval ?? 1000;
  const maxRetries = Math.max(options.maxRetries ?? 3, 0);
  const retryDelay = Math.max(options.retryDelay ?? 100, 0);

  // Resolve the log stream name
  const logStreamName =
    options.autoCreateLogStream && "logStreamNameTemplate" in options
      ? resolveLogStreamName(options.logStreamNameTemplate)
      : options.logStreamName;

  // Track created streams to avoid redundant API calls
  const createdStreams = new Set<string>();

  // Default formatter that formats message parts into a simple string
  const defaultFormatter: TextFormatter = (record) => {
    let result = "";
    for (let i = 0; i < record.message.length; i++) {
      if (i % 2 === 0) {
        result += record.message[i];
      } else {
        result += JSON.stringify(record.message[i]);
      }
    }
    return result;
  };

  const formatter = options.formatter ?? defaultFormatter;

  const logEvents: InputLogEvent[] = [];
  let currentBatchSize = 0;
  let flushTimer: ReturnType<typeof setTimeout> | null = null;
  let disposed = false;
  let flushPromise: Promise<void> | null = null;

  function scheduleFlush(): void {
    if (flushInterval <= 0 || flushTimer !== null) return;

    flushTimer = setTimeout(() => {
      flushTimer = null;
      if (logEvents.length > 0) {
        void flushEvents();
      }
    }, flushInterval);
  }

  async function flushEvents(): Promise<void> {
    if (logEvents.length === 0 || disposed) return;

    // If there's already a flush in progress, wait for it
    if (flushPromise !== null) {
      await flushPromise;
      return;
    }

    // Start a new flush operation
    flushPromise = doFlush();
    await flushPromise;
    flushPromise = null;
  }

  async function doFlush(): Promise<void> {
    if (logEvents.length === 0 || disposed) return;

    const events = logEvents.splice(0);
    currentBatchSize = 0;

    if (flushTimer !== null) {
      clearTimeout(flushTimer);
      flushTimer = null;
    }

    // Auto-create log stream if enabled (only once per stream)
    if (options.autoCreateLogStream) {
      await ensureLogStreamExists(
        client,
        options.logGroupName,
        logStreamName,
        createdStreams,
      );
    }

    await sendEventsWithRetry(events, maxRetries);
  }

  async function sendEventsWithRetry(
    events: InputLogEvent[],
    remainingRetries: number,
  ): Promise<void> {
    try {
      const command = new PutLogEventsCommand({
        logGroupName: options.logGroupName,
        logStreamName: logStreamName,
        logEvents: events,
      });

      await client.send(command);
    } catch (error) {
      if (remainingRetries > 0) {
        // Calculate exponential backoff: base, base*2, base*4, etc.
        const attemptNumber = maxRetries - remainingRetries;
        const delay = retryDelay * Math.pow(2, attemptNumber);
        await new Promise((resolve) => setTimeout(resolve, delay));
        await sendEventsWithRetry(events, remainingRetries - 1);
      } else {
        // Log to meta logger to avoid crashing the application
        const metaLogger = getLogger(["logtape", "meta", "cloudwatch-logs"]);
        metaLogger.error(
          "Failed to send log events to CloudWatch Logs after {maxRetries} retries: {error}",
          { maxRetries, error },
        );
      }
    }
  }

  function formatLogMessage(record: LogRecord): string {
    return formatter(record);
  }

  const sink: Sink & AsyncDisposable = (record: LogRecord) => {
    if (disposed) return;

    // Skip meta logger logs to prevent infinite loops
    if (
      record.category[0] === "logtape" &&
      record.category[1] === "meta" &&
      record.category[2] === "cloudwatch-logs"
    ) {
      return;
    }

    const message = formatLogMessage(record);
    const messageBytes = new TextEncoder().encode(message).length;
    const eventSize = messageBytes + OVERHEAD_PER_EVENT;

    const logEvent: InputLogEvent = {
      timestamp: record.timestamp,
      message,
    };

    logEvents.push(logEvent);
    currentBatchSize += eventSize;

    const shouldFlushBySize = currentBatchSize > MAX_BATCH_SIZE_BYTES;
    const shouldFlushByCount = logEvents.length >= batchSize;

    if (shouldFlushBySize || shouldFlushByCount) {
      void flushEvents();
    } else {
      scheduleFlush();
    }
  };

  sink[Symbol.asyncDispose] = async () => {
    if (flushTimer !== null) {
      clearTimeout(flushTimer);
      flushTimer = null;
    }
    await flushEvents();
    disposed = true;
  };

  return sink;
}
