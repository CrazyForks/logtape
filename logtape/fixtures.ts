import type { LogRecord } from "./record.ts";

export const info: LogRecord = {
  level: "info",
  category: ["my-app", "junk"],
  message: ["Hello, ", 123, " & ", 456, "!"],
  timestamp: 1700000000000,
  properties: {},
};

export const debug: LogRecord = {
  ...info,
  level: "debug",
};

export const warning: LogRecord = {
  ...info,
  level: "warning",
};

export const error: LogRecord = {
  ...info,
  level: "error",
};

export const fatal: LogRecord = {
  ...info,
  level: "fatal",
};
