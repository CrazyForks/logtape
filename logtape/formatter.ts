import type { LogLevel } from "./level.ts";
import type { LogRecord } from "./record.ts";

/**
 * A text formatter is a function that accepts a log record and returns
 * a string.
 *
 * @param record The log record to format.
 * @returns The formatted log record.
 */
export type TextFormatter = (record: LogRecord) => string;

/**
 * The severity level abbreviations.
 */
const levelAbbreviations: Record<LogLevel, string> = {
  "debug": "DBG",
  "info": "INF",
  "warning": "WRN",
  "error": "ERR",
  "fatal": "FTL",
};

/**
 * A platform-specific inspect function.  In Deno, this is {@link Deno.inspect},
 * and in Node.js/Bun it is {@link util.inspect}.  If neither is available, it
 * falls back to {@link JSON.stringify}.
 *
 * @param value The value to inspect.
 * @returns The string representation of the value.
 */
const inspect: (value: unknown) => string = eval(`(
  "Deno" in globalThis && "inspect" in globalThis.Deno &&
    typeof globalThis.Deno.inspect === "function"
    ? globalThis.Deno.inspect
    : "util" in globalThis && "inspect" in globalThis.util &&
        globalThis.util.inspect === "function"
    ? globalThis.util.inspect
    : JSON.stringify
)`);

/**
 * The default text formatter.  This formatter formats log records as follows:
 *
 * ```
 * 2023-11-14 22:13:20.000 +00:00 [INF] categoryÂ·subcategory: Hello, world!
 * ```
 *
 * @param record The log record to format.
 * @returns The formatted log record.
 */
export function defaultTextFormatter(record: LogRecord): string {
  const ts = new Date(record.timestamp);
  let msg = "";
  for (let i = 0; i < record.message.length; i++) {
    if (i % 2 === 0) msg += record.message[i];
    else msg += inspect(record.message[i]);
  }
  const category = record.category.join("\xb7");
  return `${ts.toISOString().replace("T", " ").replace("Z", " +00:00")} [${
    levelAbbreviations[record.level]
  }] ${category}: ${msg}\n`;
}

/**
 * A console formatter is a function that accepts a log record and returns
 * an array of arguments to pass to {@link console.log}.
 *
 * @param record The log record to format.
 * @returns The formatted log record, as an array of arguments for
 *          {@link console.log}.
 */
export type ConsoleFormatter = (record: LogRecord) => readonly unknown[];

/**
 * The styles for the log level in the console.
 */
const logLevelStyles: Record<LogLevel, string> = {
  "debug": "background-color: gray; color: white;",
  "info": "background-color: white; color: black;",
  "warning": "background-color: orange; color: black;",
  "error": "background-color: red; color: white;",
  "fatal": "background-color: maroon; color: white;",
};

/**
 * The default console formatter.
 *
 * @param record The log record to format.
 * @returns The formatted log record, as an array of arguments for
 *          {@link console.log}.
 */
export function defaultConsoleFormatter(record: LogRecord): readonly unknown[] {
  let msg = "";
  const values: unknown[] = [];
  for (let i = 0; i < record.message.length; i++) {
    if (i % 2 === 0) msg += record.message[i];
    else {
      msg += "%o";
      values.push(record.message[i]);
    }
  }
  const date = new Date(record.timestamp);
  const time = `${date.getUTCHours().toString().padStart(2, "0")}:${
    date.getUTCMinutes().toString().padStart(2, "0")
  }:${date.getUTCSeconds().toString().padStart(2, "0")}.${
    date.getUTCMilliseconds().toString().padStart(3, "0")
  }`;
  return [
    `%c${time} %c${levelAbbreviations[record.level]}%c %c${
      record.category.join("\xb7")
    } %c${msg}`,
    "color: gray;",
    logLevelStyles[record.level],
    "background-color: default;",
    "color: gray;",
    "color: default;",
    ...values,
  ];
}
