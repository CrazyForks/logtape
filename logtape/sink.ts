import {
  type ConsoleFormatter,
  defaultConsoleFormatter,
  defaultTextFormatter,
  type TextFormatter,
} from "./formatter.ts";
import type { LogRecord } from "./record.ts";

/**
 * A sink is a function that accepts a log record and prints it somewhere.
 * Thrown exceptions will be suppressed and then logged to the meta logger,
 * a {@link Logger} with the category `["logtape", "meta"]`.  (In that case,
 * the meta log record will not be passed to the sink to avoid infinite
 * recursion.)
 *
 * @param record The log record to sink.
 */
export type Sink = (record: LogRecord) => void;

/**
 * A factory that returns a sink that writes to a {@link WritableStream}.
 *
 * Note that the `stream` is of Web Streams API, which is different from
 * Node.js streams.  You can convert a Node.js stream to a Web Streams API
 * stream using [`stream.Writable.toWeb()`] method.
 *
 * [`stream.Writable.toWeb()`]: https://nodejs.org/api/stream.html#streamwritabletowebstreamwritable
 *
 * @example Sink to the standard error in Deno
 * ```typescript
 * const stderrSink = getStreamSink(Deno.stderr.writable);
 * ```
 *
 * @example Sink to the standard error in Node.js
 * ```typescript
 * import stream from "node:stream";
 * const stderrSink = getStreamSink(stream.Writable.toWeb(process.stderr));
 * ```
 *
 * @param stream The stream to write to.
 * @param formatter The text formatter to use.  Defaults to
 *                  {@link defaultTextFormatter}.
 * @param encoder The text encoder to use.  Defaults to an instance of
 *                {@link TextEncoder}.
 * @returns A sink that writes to the stream.
 */
export function getStreamSink(
  stream: WritableStream,
  formatter: TextFormatter = defaultTextFormatter,
  encoder: { encode(text: string): Uint8Array } = new TextEncoder(),
): Sink {
  const writer = stream.getWriter();
  return (record: LogRecord) => {
    const bytes = encoder.encode(formatter(record));
    writer.ready.then(() => writer.write(bytes));
  };
}

/**
 * A console sink factory that returns a sink that logs to the console.
 *
 * @param formatter A console formatter.  Defaults to
 *                  {@link defaultConsoleFormatter}.
 * @param console The console to log to.  Defaults to {@link console}.
 * @returns A sink that logs to the console.
 */
export function getConsoleSink(
  formatter: ConsoleFormatter = defaultConsoleFormatter,
  console: Console = globalThis.console,
): Sink {
  return (record: LogRecord) => {
    const args = formatter(record);
    if (record.level === "debug") console.debug(...args);
    else if (record.level === "info") console.info(...args);
    else if (record.level === "warning") console.warn(...args);
    else if (record.level === "error" || record.level === "fatal") {
      console.error(...args);
    } else throw new TypeError(`Invalid log level: ${record.level}.`);
  };
}
