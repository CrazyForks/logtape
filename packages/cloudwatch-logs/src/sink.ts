import {
  CloudWatchLogsClient,
  type InputLogEvent,
  PutLogEventsCommand,
} from "@aws-sdk/client-cloudwatch-logs";
import {
  getLogger,
  type LogRecord,
  type Sink,
  type TextFormatter,
} from "@logtape/logtape";
import type { CloudWatchLogsSinkOptions } from "./types.ts";

// AWS CloudWatch Logs PutLogEvents API limits
// See: https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch_limits_cwl.html
const MAX_BATCH_SIZE_EVENTS = 10000; // Maximum 10,000 events per batch
const MAX_BATCH_SIZE_BYTES = 1048576; // Maximum batch size: 1 MiB (1,048,576 bytes)
const OVERHEAD_PER_EVENT = 26; // AWS overhead per log event: 26 bytes per event

/**
 * Gets a CloudWatch Logs sink that sends log records to AWS CloudWatch Logs.
 *
 * @param options Configuration options for the CloudWatch Logs sink.
 * @returns A sink that sends log records to CloudWatch Logs.
 * @since 1.0.0
 */
export function getCloudWatchLogsSink(
  options: CloudWatchLogsSinkOptions,
): Sink & AsyncDisposable {
  const client = options.client ??
    new CloudWatchLogsClient({
      region: options.region ?? "us-east-1",
      credentials: options.credentials,
    });

  const batchSize = Math.min(
    Math.max(options.batchSize ?? 1000, 1),
    MAX_BATCH_SIZE_EVENTS,
  );
  const flushInterval = options.flushInterval ?? 1000;
  const maxRetries = Math.max(options.maxRetries ?? 3, 0);
  const retryDelay = Math.max(options.retryDelay ?? 100, 0);

  // Default formatter that formats message parts into a simple string
  const defaultFormatter: TextFormatter = (record) => {
    let result = "";
    for (let i = 0; i < record.message.length; i++) {
      if (i % 2 === 0) {
        result += record.message[i];
      } else {
        result += JSON.stringify(record.message[i]);
      }
    }
    return result;
  };

  const formatter = options.formatter ?? defaultFormatter;

  const logEvents: InputLogEvent[] = [];
  let currentBatchSize = 0;
  let flushTimer: ReturnType<typeof setTimeout> | null = null;
  let disposed = false;

  function scheduleFlush(): void {
    if (flushInterval <= 0 || flushTimer !== null) return;

    flushTimer = setTimeout(() => {
      flushTimer = null;
      if (logEvents.length > 0) {
        void flushEvents();
      }
    }, flushInterval);
  }

  async function flushEvents(): Promise<void> {
    if (logEvents.length === 0 || disposed) return;

    const events = logEvents.splice(0);
    currentBatchSize = 0;

    if (flushTimer !== null) {
      clearTimeout(flushTimer);
      flushTimer = null;
    }

    await sendEventsWithRetry(events, maxRetries);
  }

  async function sendEventsWithRetry(
    events: InputLogEvent[],
    remainingRetries: number,
  ): Promise<void> {
    try {
      const command = new PutLogEventsCommand({
        logGroupName: options.logGroupName,
        logStreamName: options.logStreamName,
        logEvents: events,
      });

      await client.send(command);
    } catch (error) {
      if (remainingRetries > 0) {
        // Calculate exponential backoff: base, base*2, base*4, etc.
        const attemptNumber = maxRetries - remainingRetries;
        const delay = retryDelay * Math.pow(2, attemptNumber);
        await new Promise((resolve) => setTimeout(resolve, delay));
        await sendEventsWithRetry(events, remainingRetries - 1);
      } else {
        // Log to meta logger to avoid crashing the application
        const metaLogger = getLogger(["logtape", "meta", "cloudwatch-logs"]);
        metaLogger.error(
          "Failed to send log events to CloudWatch Logs after {maxRetries} retries: {error}",
          { maxRetries, error },
        );
      }
    }
  }

  function formatLogMessage(record: LogRecord): string {
    return formatter(record);
  }

  const sink: Sink & AsyncDisposable = (record: LogRecord) => {
    if (disposed) return;

    // Skip meta logger logs to prevent infinite loops
    if (
      record.category[0] === "logtape" &&
      record.category[1] === "meta" &&
      record.category[2] === "cloudwatch-logs"
    ) {
      return;
    }

    const message = formatLogMessage(record);
    const messageBytes = new TextEncoder().encode(message).length;
    const eventSize = messageBytes + OVERHEAD_PER_EVENT;

    const logEvent: InputLogEvent = {
      timestamp: record.timestamp,
      message,
    };

    logEvents.push(logEvent);
    currentBatchSize += eventSize;

    const shouldFlushBySize = currentBatchSize > MAX_BATCH_SIZE_BYTES;
    const shouldFlushByCount = logEvents.length >= batchSize;

    if (shouldFlushBySize || shouldFlushByCount) {
      void flushEvents();
    } else {
      scheduleFlush();
    }
  };

  sink[Symbol.asyncDispose] = async () => {
    if (flushTimer !== null) {
      clearTimeout(flushTimer);
      flushTimer = null;
    }
    await flushEvents();
    disposed = true;
  };

  return sink;
}
