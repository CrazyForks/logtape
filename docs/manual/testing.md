Testing
=======

Here are some tips for testing your application or library with LogTape.


Reset configuration
-------------------

You can reset the configuration of LogTape to its initial state.  This is
useful when you want to reset the configuration between tests.  For example,
the following code shows how to reset the configuration after a test
(regardless of whether the test passes or fails) in Deno:

~~~~ typescript
import { configure, reset } from "@logtape/logtape";

Deno.test("my test", async (t) => {
  await t.step("set up", async () => {
    await configure({ /* ... */ });
  });

  await t.step("run test", () => {
    // Run the test
  });

  await t.step("tear down", async () => {
    await reset();  // [!code highlight]
  });
});
~~~~


Buffer sink
-----------

For testing purposes, you may want to collect log messages in memory.  Although
LogTape does not provide a built-in buffer sink, you can easily implement it:

~~~~ typescript
import { type LogRecord, configure } from "@logtape/logtape";

const buffer: LogRecord[] = [];

await configure({
  sinks: {
    buffer: buffer.push.bind(buffer),  // [!code highlight]
  },
  // Omitted for brevity
});
~~~~
