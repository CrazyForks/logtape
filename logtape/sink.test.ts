import { assertThrows } from "@std/assert";
import { delay } from "@std/async";
import { assertSnapshot } from "@std/testing/snapshot";
import makeConsoleMock from "consolemock";
import { debug, error, fatal, info, warning } from "./fixtures.ts";
import { defaultConsoleFormatter } from "./formatter.ts";
import type { LogLevel } from "./record.ts";
import { getConsoleSink, getStreamSink } from "./sink.ts";

interface ConsoleMock extends Console {
  history(): unknown[];
}

Deno.test("getStreamSink()", async (t) => {
  let buffer: string = "";
  const decoder = new TextDecoder();
  const sink = getStreamSink(
    new WritableStream({
      write(chunk) {
        buffer += decoder.decode(chunk);
        return Promise.resolve();
      },
    }),
  );
  sink(debug);
  sink(info);
  sink(warning);
  sink(error);
  sink(fatal);
  await delay(100);
  await assertSnapshot(t, buffer);
});

Deno.test("getConsoleSink()", async (t) => {
  // @ts-ignore: consolemock is not typed
  const mock: ConsoleMock = makeConsoleMock();
  const sink = getConsoleSink(defaultConsoleFormatter, mock);
  sink(debug);
  sink(info);
  sink(warning);
  sink(error);
  sink(fatal);
  await assertSnapshot(t, mock.history());

  assertThrows(
    () => sink({ ...info, level: "invalid" as LogLevel }),
    TypeError,
    "Invalid log level: invalid.",
  );
});
