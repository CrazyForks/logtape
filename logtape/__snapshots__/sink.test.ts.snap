export const snapshot = {};

snapshot[`getStreamSink() 1`] = `
"2023-11-14 22:13:20.000 +00:00 [DBG] Hello, 123 & 456!
2023-11-14 22:13:20.000 +00:00 [INF] Hello, 123 & 456!
2023-11-14 22:13:20.000 +00:00 [WRN] Hello, 123 & 456!
2023-11-14 22:13:20.000 +00:00 [ERR] Hello, 123 & 456!
2023-11-14 22:13:20.000 +00:00 [FTL] Hello, 123 & 456!
"
`;

snapshot[`getConsoleSink() 1`] = `
[
  {
    DEBUG: [
      "%cDEBUG%c %cmy-app·junk %cHello, %o & %o!",
      "background-color: gray; color: white;",
      "background-color: default;",
      "color: gray;",
      "color: default;",
      123,
      456,
    ],
  },
  {
    INFO: [
      "%cINFO%c %cmy-app·junk %cHello, %o & %o!",
      "background-color: white; color: black;",
      "background-color: default;",
      "color: gray;",
      "color: default;",
      123,
      456,
    ],
  },
  {
    WARN: [
      "%cWARNING%c %cmy-app·junk %cHello, %o & %o!",
      "background-color: orange;",
      "background-color: default;",
      "color: gray;",
      "color: default;",
      123,
      456,
    ],
  },
  {
    ERROR: [
      "%cERROR%c %cmy-app·junk %cHello, %o & %o!",
      "background-color: red;",
      "background-color: default;",
      "color: gray;",
      "color: default;",
      123,
      456,
    ],
  },
  {
    ERROR: [
      "%cFATAL%c %cmy-app·junk %cHello, %o & %o!",
      "background-color: maroon;",
      "background-color: default;",
      "color: gray;",
      "color: default;",
      123,
      456,
    ],
  },
]
`;
